NPM package for installing faas-cli

doowb opened the pull request:
Signed-off-by: Brian Woodward <brian.woodward@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This PR introduces an `npm` package that, when installed, will run a `postinstall` script to download and install the latest version of the `faas-cli`.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Related to issue #645 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Still a work in progress and only tested on Mac OSX (darwin).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

alexellis commented:
This looks very promising @doowb - have you got an ETA for landing this?

doowb commented:
My wedding was a couple of days ago, so last month was hectic. Now that's passed, I'll make some time this week for this PR. One blocker that I think was mentioned in #645 was getting the install script to detect when `npm install` was run with `sudo`.

alexellis commented:
Congratulations!

I did not know that you were away. Can you give an estimate by when you expect to finish the work? An additional 1, 2, 4, 8 weeks?

doowb commented:
> Congratulations!

Thanks!

Sorry for the delay... I just pushed up changes that I think will work on all OS's. I'll be trying to test this on Windows tonight and report back. If anyone else can test on different systems, that would help a lot.

burtonr commented:
Just tested this on Windows 10 (Powershell):

```
PS C:\Users\burtonr\source\repos\faas-cli\npm> npm install

> openfaas-cli@1.0.0 postinstall C:\Users\burtonr\source\repos\faas-cli\npm
> node install.js

Downloading package https://github.com/openfaas/faas-cli/releases/download/0.8.21/faas-cli.exe to C:\Users\burtonr\.openfaas\bin\faas-cli.exe
{ Error: Command failed: chmod +x C:\Users\burtonr\.openfaas\bin\faas-cli.exe
'chmod' is not recognized as an internal or external command,
operable program or batch file.

    at ChildProcess.exithandler (child_process.js:294:12)
    at ChildProcess.emit (events.js:198:13)
    at maybeClose (internal/child_process.js:982:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
  killed: false,
  code: 1,
  signal: null,
  cmd: 'chmod +x C:\\Users\\burtonr\\.openfaas\\bin\\faas-cli.exe',
  stdout: '',
  stderr:
   '\'chmod\' is not recognized as an internal or external command,\r\noperable program or batch file.\r\n' }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! openfaas-cli@1.0.0 postinstall: `node install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the openfaas-cli@1.0.0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\burtonr\AppData\Roaming\npm-cache\_logs\2019-07-20T07_28_04_814Z-debug.log
PS C:\Users\burtonr\source\repos\faas-cli\npm>
```

It seems on Windows, that is not necessary, so you could simply add this little `if` block:
```
  if (extname !== '.exe') {
    await cmd(`chmod +x ${dest}`);
  }
```

Also, I can't seem to get the `bin` to work as it should. Not sure if I'm missing something on my setup though. It's a fresh Windows machine with Node.js LTS freshly installed as well. No other configuration done.

On a more positive note, I also tested on Ubuntu Linux and it worked exactly as expected.

I'm not sure how I feel about having it install in a different directory than the standard CLI bash install, but I can see why it's done that way. Using the `bin` scripts makes it transparent to the user, so I'm not totally against it.

I think once you get that `chmod` updated to be Windows compatible, and the `bin` issue, we should be good to merge this and start getting some testers to try it out (me included!)

doowb commented:
Thanks for testing this. I'll make the update and try it on Windows to see if I can figure out the issues.

> I'm not sure how I feel about having it install in a different directory than the standard CLI bash install, but I can see why it's done that way. Using the `bin` scripts makes it transparent to the user, so I'm not totally against it.

I was hesitant to do this at first, but I think this is the only way to ensure npm installs it correctly (mainly because I wasn't able to detect if the user was `root` or not from the install script).

doowb commented:
@burtonr I updated the install script so it doesn't use `chmod` on WIndows.
I tested this on a Windows machine in a normal command line (`cmd`), a git bash command line (same as `cmd` but with some extra git tools added, and PowerShell and it all worked fine.

One thing I noticed above is that you ran `npm install` from the `faas-cli/npm` directory. For some reason npm will run the `postinstall` script in that case, but it doesn't actually install the binaries. Normally when you install something with a `bin` property it will install the module in the "local" `node_modules` folder and the cli will be in `node_modules/.bin`. This lets users use npm scripts to access the bin locally.

The way I'm testing is using npm's feature to install modules from a file path. If I create another npm project next to `faas-cli`, I can install it by doing `npm install ../faas-cli/npm` and it will go into my `./node_modules/.bin` folder. Npm scripts can then use `faas` by adding a script to the `package.json` like this:

```json
{
  "scripts": {
    "faas": "faas",
    "build": "faas build",
    "push": "faas push",
    "deploy": "faas deploy",
    "up": "faas up"
  }
}
```

Then `faas` can be used in the local project like: `npm run faas template store list` or `npm run up`.

This can be useful for CI, but on my machine I like installing globally:

```sh
npm install --global ../faas-cli/npm
```

I hope this helps... you might have already known all of this but I thought it would be useful for people that see this and don't know. I'll update the `README.md` in the `faas-cli/npm` folder to include some of this information since that would be viewable on the `npmjs.com` site.

doowb commented:
> I'm not sure how I feel about having it install in a different directory than the standard CLI bash install

This got me thinking... what if we just install it into a folder relative to where npm is installing the rest of the module. This will have the added benefit of removing the binary if someone does `npm uninstall --global openfaas-cli`

burtonr commented:
Looking great now! Just tried again on both Linux and Windows (using the `-g` option) and worked perfectly!

I would agree that the binary installation should be more in line with the "npm way" and tuck it in with the rest of the package. Like you said, it will make uninstall easier, but also it will be easier to find, or more familiar to those who are used to installing tools via npm. They'll know where to look if they need to.

I think once that's good, I'm good with this! :100: Nice work!

alexellis commented:
Please remember to remove the WIP tag, when ready for review/merge

doowb commented:
BTW... I'm going to make some time to update the README.md in the `npm` folder tomorrow night with information I put here and #645. Other than that, I think this PR is ready for review.

alexellis commented:
How do we publish this package and when?

- for the `npm` namespace / account (I have `openfaas` as a user, not an org and couldn't convert it)
- when do we update it? every time we cut a release tag?

Are we determining the version to download using the same mechanism as the GET.sh script? We moved away from using GitHub's API because it gets rate-limited very aggressively 

Alex

doowb commented:
> I'm not sure if that can be automatic/dynamic, but for now, let's just update it to the current release version `0.8.21` to line everything up.

I agree with this since it will allow people to continue to use `npm` to update their installed version.

> How do we publish this package and when?

- The `@openfaas` namespace would work. With it being a user, you'd just have to manage the "owners" manually like you do with other npm packages. The name could be `@openfaas/cli` or `@openfaas/faas-cli` to match the GitHub repository.
- I think this should be updated whenever the cli executable is updated. I'm not sure if the cli does this already, but there are ways to detect a new version is published to npm and notify the user on how to update their version. If this is done, I think it should be done first so users are always notified.

> Are we determining the version to download using the same mechanism as the GET.sh script? We moved away from using GitHub's API because it gets rate-limited very aggressively

This install script is using the GitHub API, but it's only making 2 anonymous calls (get the latest release, get the assets for that release).

burtonr commented:
Hey @doowb I think we're almost there! 3 minor things though...

1) Could you fix the version to match the faas-cli release number?
It will need to be a manual process for the time being to update the package.json when we create a new version. I think we could get a bash script to do it on build/deploy with Travis. Not now, but soon after.
2) Use the version number to pull the specific version from Gtihub's Releases.
This will eliminate 1 of the calls to Github, and ensure that the npm version gets the same version of the CLI. Wouldn't it be a crazy (bad) surprise to have version 0.8.7 of the npm package download version 0.9.1 of the cli? (as an example)
3) Update the name to `faas-cli` so that the install will be `@openfaas/faas-cli` 
This will match the repo, the binary, and the brew names. Be sure to update the references in both the readme files as well.

doowb commented:
@burtonr thanks for reviewing. I'll make these updates by the end of the week.

doowb commented:
> 1. Could you fix the version to match the faas-cli release number?

Updated `package.json` to use version `0.9.1` which is the current release.

> 2. Use the version number to pull the specific version from Gtihub's Releases.

Using the `.version` from `package.json` to get the specific versioned/tagged release. I also noticed that the "assets" are all returned in that payload so there was never a need for the second call to the GitHub api.

> 3. Update the name to `faas-cli` so that the install will be `@openfaas/faas-cli`

I updated the name to `@openfaas/faas-cli`. The name `faas-cli` is taken by someone else on NPM. Either way the binaries are aliased as `faas-cli` and `faas` in the terminal so the user experience is the same as if they installed the cli another way.

I think this is good to go. Let me know if you see anything else (I also sync'd changes from `master` in it mattered for the README.

/cc @burtonr

doowb commented:
I screwed up merging master into this branch and forgot to sign the commit. I squashed all the commits to clean it up.

burtonr commented on README.md:
````diff
@@ -39,6 +39,14 @@ $ brew install faas-cli
+Via npm:
+
+```
+$ npm install --global openfaas-cli
````
This should be updated to reflect the updated name `@openfaas/faas-cli`

burtonr commented on npm/README.md:
````diff
@@ -0,0 +1,65 @@
+Install with [npm](https://www.npmjs.com/) as a global module:
+
+```sh
+$ npm install --global openfaas-cli
````
again, update the command with the updated name `@openfaas/faas-cli`

burtonr commented on npm/README.md:
````diff
@@ -0,0 +1,65 @@
+Install with [npm](https://www.npmjs.com/) as a local dev dependency:
+
+```sh
+$ npm install --save-dev openfaas-cli
````
updated name: `@openfaas/faas-cli`

doowb commented:
@burtonr done! Thanks for catching the README changes.

doowb commented:
> I'm happy to approve this as is with the knowledge that there is some work ahead to make the publishing and versioning scripted.

I'll spend some time tomorrow on making sure this is up-to-date with the latest version and looking at the CI process (if there is one), or create something simple if there isn't. I'm happy to try to submit that as a separate PR or include it here.

alexellis commented:
Thank you

doowb commented:
I've updated the version to `0.9.3` in the `npm/package.json` file.
It looks like the releases are done through travis-ci. I'm not sure how to get the version from that to update `npm/package.json`, but it's something I can research later and implement in another PR, if that's okay.

rzr commented:
git describe --tags could help but you'll need to tweak this to match semver convention used in npm...

alexellis commented on npm/README.md:
```diff
@@ -0,0 +1,65 @@
+
+> OpenFaaS CLI
+
+## Install globally
```
Let's just state "global" install. I can't think of a reason to use it in any other way?

doowb replied:
I'm good with that.

alexellis commented on npm/install.js:
```diff
@@ -0,0 +1,94 @@
+'use strict';
```
This would perhaps need to be included in CI in some way for an e2e test.

alexellis commented on README.md:
````diff
@@ -39,6 +39,14 @@ $ brew install faas-cli
+Via npm:
+
+```
+$ npm install --global @openfaas/faas-cli
````
what about `-g` that is more common?

Who owns `@openfaas`, is that the account I have? How are we pushing to it and auth'ing to npm during CI for this?

doowb replied:
`@openfaas` is a [user](https://www.npmjs.com/~openfaas) on NPM. I thought it was an org already. There are ways to auth an push during CI, but I'm not positive on how that's done.

For the `-g`, I usually spell out the entire word in documentation so there isn't any confusion over what it stands for. I'm open to making it just `-g` though.

burtonr replied:
This would be a use for adding a `.npmrc` file to the repo (in the same directory as the `package.json` file)
There, you would add the auth token that you would need to generate locally first. Of course, you would want to keep that in a (TravisCI secret](https://docs.travis-ci.com/user/deployment-v2#sts=Securing%20secrets%20#)

My thoughts:
* Run `npm login` locally for the `@openfaas` user
* Get the auth token generated: `cat ~/.npmrc`
* Use Travis CLI to encrypt that token: `travis encrypt <auth token>`
* Add that encrypted value to the travis build's environment
* Add the `.npmrc` file and reference that env var 
Examples here: https://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules

alexellis commented:
Added to a milestone, PTAL @doowb ?

alexellis commented:
I know this PR is not merged, but perhaps if we get it merged it (and mark it as not-ready), then it will motivate someone else to finish the work?

burtonr commented:
Happy to take this through the next steps! I'll go through each of your comments in a new PR soon! :+1:

burtonr commented:
FYI @alexellis and anyone else, the updates and testing are available in this PR: https://github.com/openfaas/faas-cli/pull/723
